firmware/application/external/rook/ui_rook.cpp

```diff
@@ -0,0 +1,269 @@
+        transmitter_model.set_baseband_bandwidth(1750000);              // Set the baseband bandwidth
+        transmitter_model.enable();                                     // Enable the transmitter
+ 
+        // Configure OOK data and transmission characteristics
+        baseband::set_ook_data(
+            bitstream_length,                 // Length of the bitstream to transmit
+            SAMPLE_RATE_VALUE / SYMBOL_RATE_VALUE, // Calculate symbol period based on repetition rate
+            REPEAT,                           // Set the number of repetitions per symbol
+            PAUSE_SYMBOL                      // Set the pause between symbols
+        );
+    }
+ 
+    // `stop_tx` method: Stops the transmission and resets the progress bar.
+    void ROOKAppView::stop_tx() {
+        transmitter_model.disable();                   // Disable the transmitter
+        progressBar_progress.set_value(0);             // Reset progress bar to 0
+    }
+ 
+    // `on_file_changed` method: Called when a new file is loaded; parses file data into variables
+    void ROOKAppView::on_file_changed(const fs::path& new_file_path) {
+        payload.clear();  // Clear previous payload content
+ 
+        File data_file;
+        auto open_result = data_file.open(new_file_path);
+        if (open_result) {
+            nav_.display_modal("Error", "Unable to open file.");
+            return;
+        }
+ 
+        FileLineReader reader(data_file);
+        for (const auto& line : reader) {
+            // Split the line into segments to extract each value
+            size_t first_space = line.find(' ');
+            size_t second_space = line.find(' ', first_space + 1);
+            size_t third_space = line.find(' ', second_space + 1);
+            size_t fourth_space = line.find(' ', third_space + 1);
+            size_t fifth_space = line.find(' ', fourth_space + 1);
+ 
+            // Extract each component of the line
+            std::string frequency_str = line.substr(0, first_space);
+            std::string sample_rate_str = line.substr(first_space + 1, second_space - first_space - 1);
+            std::string symbols_rate_str = line.substr(second_space + 1, third_space - second_space - 1);
+            std::string repeat_str = line.substr(third_space + 1, fourth_space - third_space - 1);
+            std::string pause_symbol_str = line.substr(fourth_space + 1, fifth_space - fourth_space - 1);
+            std::string payload_data = line.substr(fifth_space + 1);  // Extract binary payload as final value
+ 
+            // Convert and assign frequency
+            unsigned long long frequency = std::stoull(frequency_str);
+            field_frequency.set_value(frequency);
+            transmitter_model.set_target_frequency(frequency);
+ 
+            // Convert and assign symbols rate
+            unsigned int symbols_rate = std::stoi(symbols_rate_str);
```

Comment:
we gave up using stoi in this project to save some spaces and resolve some potential performance issue.
I recommend using atoi

This may need to fixed unless @htotoo agreed

Reply:
Since no error checks implemented, i don't think it is the time, where we need to include this.to the fw base

Reply:
sorry i tried but i don't understand what you meant

Reply:
Oh i got it, you meant to handle the situation that user loaded a wrong file?
I agree

Reply:
He meant handling the outcomes from a bad string to int conversion, catching the exception and do something about it.
If it's not planned to handle exceptions, then atoi() is enough.

Reply:
Replaced stoi with atoi for conversion